helpers/utils.go

```diff
@@ -18,3 +23,32 @@ func ContainsCommand(commands []commands.Command, command commands.Command) bool
 	}
 	return false
 }
+
+// GetFreePort asks the kernel for a free port.
+func GetFreePort() (string, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return "", err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return "", err
+	}
+	defer l.Close()
+
+	return strconv.Itoa(l.Addr().(*net.TCPAddr).Port), nil
+}
+
+// WaitForPort waits for a port to be open.
+func WaitForPort(port string) error {
+	for i := 0; i < 10; i++ {
+		conn, err := net.DialTimeout("tcp", "localhost:"+port, 500*time.Millisecond)
+		if err == nil {
+			_ = conn.Close()
+			return nil
+		}
+		time.Sleep(500 * time.Second)
+	}
+	return errors.New("Timeout waiting for port " + port)
+}
```

Comment:
```suggestion
func WaitForPort(ctx context.Context, port string) error {
	for i := 0; i < 10; i++ {
		conn, err := net.DialTimeout("tcp", "localhost:"+port, 500*time.Millisecond)
		if err == nil {
			_ = conn.Close()
			return nil
		}

		select {
			case <- time.After(500 * time.Second):
			case <- ctx.Done():
				return ctx.Err()
		}
	}
	return errors.New("Timeout waiting for port " + port)
}
```

Reply:
While I doubt if this can ever have a blocking effect on the app (this is isn't called from the main thread) I agree its the more "correct" way to do this.

Even more correct would be to pass the context to the dialer as well 😏 But again, this code will never be blocking the app so doubt if it adds any value.

But I can have a quick look tomorrow and add that last bit.